File: Python/2365.py
import sys
import copy
from collections import deque


def init(x):
    res = copy.deepcopy(base)
    _cur = N
    for _i in range(1, N + 1):
        for _j in range(last - N, last):
            _cur += 1
            res[_i][_cur] = x
    return res


N = int(sys.stdin.readline())
row = list(map(int, sys.stdin.readline().split()))
column = list(map(int, sys.stdin.readline().split()))
length = N ** 2 + 2 * N + 2
last = length - 1
base = [[0] * length for _ in range(length)]
network = [[] for _ in range(length)]
for i in range(1, N + 1):
    network[0].append(i)
    network[i].append(0)
    base[0][i] = row[i - 1]
    network[last].append(last - i)
    network[last - i].append(last)
    base[last - i][last] = column[-i]
cur = N
for i in range(1, N + 1):
    for j in range(last - N, last):
        cur += 1
        network[i].append(cur)
        network[cur].append(i)
        network[cur].append(j)
        network[j].append(cur)
        base[cur][j] = column[j - last]
# print(network)
# print(base)
tmpsum = sum(column)
left = 0
right = max(max(row), max(column))
ans = right
while right >= left:
    mid = (left + right) // 2
    flow = init(mid)
    tmp = 0
    while True:
        q = deque()
        q.append(0)
        back = [-1] * length
        while q:
            u = q.popleft()
            for v in network[u]:
                if back[v] == -1 and flow[u][v]:
                    back[v] = u
                    q.append(v)
        if back[last] == -1:
            break
        cur = last
        val = sys.maxsize
        while cur:
            val = min(val, flow[back[cur]][cur])
            cur = back[cur]
        cur = last
        while cur:
            flow[back[cur]][cur] -= val
            flow[cur][back[cur]] += val
            cur = back[cur]
        tmp += val
    if tmp != tmpsum:
        left = mid + 1
    else:
        right = mid - 1
        ans = mid
print(ans)
flow = init(ans)
while True:
    q = deque()
    q.append(0)
    back = [-1] * length
    while q:
        u = q.popleft()
        for v in network[u]:
            if back[v] == -1 and flow[u][v]:
                back[v] = u
                q.append(v)
    if back[last] == -1:
        break
    cur = last
    val = sys.maxsize
    while cur:
        val = min(val, flow[back[cur]][cur])
        cur = back[cur]
    cur = last
    while cur:
        flow[back[cur]][cur] -= val
        flow[cur][back[cur]] += val
        cur = back[cur]
for i in range(N + 1, N ** 2 + N + 1, N):
    tmp = (i - 1) // N
    for j in range(i, i + N):
        print(flow[j][tmp], end=' ')
    print()

File: C++/2365.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <cstring>
#include <queue>
#include <climits>
using namespace std;

int N, row[50], column[50], length, last, base[2602][2602], flow[2602][2602], l, r, mid, ans, tmpsum, val, cur, pre[2602];
vector<int> network[2602];

void init(int x) {
	memcpy(flow, base, sizeof(base));
	cur = N;
	for (int i = 1; i <= N; i++) {
		for (int j = last - N; j < last; j++) {
			cur++;
			flow[i][cur] = x;
		}
	}
}


int main() {
	cin.tie(NULL);
	cout.tie(NULL);
	ios::sync_with_stdio(false);
	cin >> N;
	for (int i = 0; i < N; i++) {
		cin >> row[i];
		tmpsum += row[i];
		r = max(r, row[i]);
	}
	for (int i = 0; i < N; i++) {
		cin >> column[i];
		r = max(r, column[i]);
	}
	length = N * N + 2 * N + 2;
	last = length - 1;
	for (int i = 1; i <= N; i++) {
		network[0].push_back(i);
		network[i].push_back(0);
		base[0][i] = row[i - 1];
		network[last].push_back(last - i);
		network[last - i].push_back(last);
		base[last - i][last] = column[N - i];
	}
	cur = N;
	for (int i = 1; i <= N; i++) {
		for (int j = last - N; j < last; j++) {
			cur++;
			network[i].push_back(cur);
			network[cur].push_back(i);
			network[cur].push_back(j);
			network[j].push_back(cur);
			base[cur][j] = column[N + j - last];
		}
	}
	ans = r;
	while (r >= l) {
		mid = (l + r) / 2;
		init(mid);
		int tmp = 0;
		while (true) {
			queue<int> q;
			q.push(0);
			memset(pre, -1, sizeof(pre));
			while (!q.empty()) {
				int u = q.front();
				q.pop();
				for (int v : network[u]) {
					if (pre[v] == -1 && flow[u][v]) {
						pre[v] = u;
						q.push(v);
					}
				}
			}
			if (pre[last] == -1)
				break;
			cur = last;
			val = INT_MAX;
			while (cur) {
				val = min(val, flow[pre[cur]][cur]);
				cur = pre[cur];
			}
			cur = last;
			while (cur) {
				flow[pre[cur]][cur] -= val;
				flow[cur][pre[cur]] += val;
				cur = pre[cur];
			}
			tmp += val;
		}
		if (tmp == tmpsum) {
			r = mid - 1;
			ans = mid;
		}
		else
			l = mid + 1;
	}
	cout << ans << '\n';
	init(ans);
	while (true) {
		queue<int> q;
		q.push(0);
		memset(pre, -1, sizeof(pre));
		while (!q.empty()) {
			int u = q.front();
			q.pop();
			for (int v : network[u]) {
				if (pre[v] == -1 && flow[u][v]) {
					pre[v] = u;
					q.push(v);
				}
			}
		}
		if (pre[last] == -1)
			break;
		cur = last;
		val = INT_MAX;
		while (cur) {
			val = min(val, flow[pre[cur]][cur]);
			cur = pre[cur];
		}
		cur = last;
		while (cur) {
			flow[pre[cur]][cur] -= val;
			flow[cur][pre[cur]] += val;
			cur = pre[cur];
		}
	}
	for (int i = N + 1; i <= N * N + N; i += N) {
		int tmp = (i - 1) / N;
		for (int j = i; j < i + N; j++) {
			cout << flow[j][tmp] << ' ';
		}
		cout << '\n';
	}
}
